File: rxkata/src/main/java/org/sergiiz/rxkata/CountriesServiceSolved.java
package org.sergiiz.rxkata;

import java.util.List;
import java.util.Map;
import java.util.concurrent.FutureTask;

import io.reactivex.Observable;
import io.reactivex.Single;

class CountriesServiceSolved implements CountriesService {

    @Override
    public Single<String> countryNameInCapitals(Country country) {
        return null; // put your solution here
    }

    public Single<Integer> countCountries(List<Country> countries) {
        return null; // put your solution here
    }

    public Observable<Long> listPopulationOfEachCountry(List<Country> countries) {
        return null; // put your solution here;
    }

    @Override
    public Observable<String> listNameOfEachCountry(List<Country> countries) {
        return null; // put your solution here
    }

    @Override
    public Observable<Country> listOnly3rdAnd4thCountry(List<Country> countries) {
        return null; // put your solution here
    }

    @Override
    public Single<Boolean> isAllCountriesPopulationMoreThanOneMillion(List<Country> countries) {
        return null; // put your solution here
    }

    @Override
    public Observable<Country> listPopulationMoreThanOneMillion(List<Country> countries) {
        return null; // put your solution here
    }

    @Override
    public Observable<Country> listPopulationMoreThanOneMillionWithTimeoutFallbackToEmpty(final FutureTask<List<Country>> countriesFromNetwork) {
        return null; // put your solution here
    }

    @Override
    public Observable<String> getCurrencyUsdIfNotFound(String countryName, List<Country> countries) {
        return null; // put your solution here
    }

    @Override
    public Observable<Long> sumPopulationOfCountries(List<Country> countries) {
        return null; // put your solution here
    }

    @Override
    public Single<Map<String, Long>> mapCountriesToNamePopulation(List<Country> countries) {
        return null; // put your solution here
    }

    @Override
    public Observable<Long> sumPopulationOfCountries(Observable<Country> countryObservable1,
                                                     Observable<Country> countryObservable2) {
        return null; // put your solution here
    }

    @Override
    public Single<Boolean> areEmittingSameSequences(Observable<Country> countryObservable1,
                                                    Observable<Country> countryObservable2) {
        return null; // put your solution here
    }

    @Override
    public Observable<Tuple<String, Long>> getAveragePopulationByCurrency(List<Country> countries) {
        return null; // put your solution here
    }
}

File: src/main/java/Rx/CountriesServiceSolved.kt
package Rx
import Rx.CountriesService
import java.util.concurrent.FutureTask

import io.reactivex.Observable
import io.reactivex.Single

class CountriesServiceSolved : CountriesService {

    override fun countryNameInCapitals(country: Country): Single<String> {
        val capName = country.name.toUpperCase()
        return Single.just(capName)
    }

     override fun countCountries(countries: List<Country>): Single<Int> {
        val count = countries.count()
        return Single.just(count)
    }

    override fun listPopulationOfEachCountry(countries: List<Country>): Observable<Long> {
        val obsCountries = Observable.fromIterable(countries)
        return  obsCountries.map { it.population }
    }

    override fun listNameOfEachCountry(countries: List<Country>): Observable<String> {
        val obsCountries = Observable.fromIterable(countries)
        return obsCountries.map { it.name }
    }

    override fun listOnly3rdAnd4thCountry(countries: List<Country>): Observable<Country> {
        val thirdandforth = listOf<Country>(countries[2],countries[3])
        return Observable.fromIterable(thirdandforth)
    }

//    override fun isAllCountriesPopulationMoreThanOneMillion(countries: List<Country>): Single<Boolean>? {
//        return null // put your solution here
//    }
//
//    override fun listPopulationMoreThanOneMillion(countries: List<Country>): Observable<Country>? {
//        return null // put your solution here
//    }
//
//    override fun listPopulationMoreThanOneMillionWithTimeoutFallbackToEmpty(countriesFromNetwork: FutureTask<List<Country>>): Observable<Country> {
//        return null // put your solution here
//    }
//
//    override fun getCurrencyUsdIfNotFound(countryName: String, countries: List<Country>): Observable<String>? {
//        return null // put your solution here
//    }
//
//    override fun sumPopulationOfCountries(countries: List<Country>): Observable<Long>? {
//        return null // put your solution here
//    }
//
//    override fun mapCountriesToNamePopulation(countries: List<Country>): Single<Map<String, Long>>? {
//        return null // put your solution here
//    }
//
//    override fun sumPopulationOfCountries(countryObservable1: Observable<Country>,
//                                          countryObservable2: Observable<Country>): Observable<Long> {
//        return null // put your solution here
//    }
//
//    override fun areEmittingSameSequences(countryObservable1: Observable<Country>,
//                                          countryObservable2: Observable<Country>): Single<Boolean> {
//        return null // put your solution here
//    }
}
